### DISCUSSION TITLE
Can I use passport-local-mongoose to authenticate users? Using next.js/with-passport

### DISCUSSION THREAD

**Discussion by drscottlobo** [vercel/next.js#25948] [Help]
I'm building on top of the nextjs/with-passport example app and I'm using passport-local-mongoose to create the salt and hash for my created users. Saving the new user with the mongoose's auto-generated salt and hash works, but when it comes time to authenticate the user, the passport method .authenticate() returns an error: 'empty password'.

Here's my user model: 

import mongoose from 'mongoose';
import passportLocalMongoose from 'passport-local-mongoose';
const Schema = mongoose.Schema;

const userSchema = new Schema({
    username: {type: String, unique: true},
}, {timestamps: true});

userSchema.plugin(passportLocalMongoose);

mongoose.models = {};

export default mongoose.model('User', userSchema);

////// And I'm trying to use passport-local-mongoose to handle the authentication at this point: 

export async function validatePassword(user, inputPassword) {
   const isAuthenticated = await user.authenticate()(user, inputPassword);
  return isAuthenticated;
}

The user and password are indeed showing up from the inputs, any ideas what I'm doing wrong? Am I calling the authenticate method incorrectly? I've also tried user.authenticate(user, inputPassowrd)...

Thanks for any suggestions!

**Comment by SolwayEnterprises** [comment]
I've had a long journey of trying to figure out myself how to use passport-local-mongoose in Next.js (though I was a novice coder and brand new to Next.js when I started trying to convert it from Express, so I guess it's no surprise it took a while). I just got it working and this really needs more documentation tips, so I'm happy to share in case you're still working on this or anyone else runs into this situation.

First, in the middleware folder, there's the setting up of Mongoose:
![dbConnect](https://user-images.githubusercontent.com/97817595/209577505-c1f56002-d057-4aea-9f81-cea0812de603.png)

Then, plugging in passport-local-mongoose to the `UserSchema` before exporting it (`./models/User.js`):
![userModel](https://user-images.githubusercontent.com/97817595/209577809-61fe0f84-e124-4f5b-9dec-ddef37d4a843.png)

And, as it sounds like you were able to do, I was able to register the user (`./pages/api/register.js`):
![register](https://user-images.githubusercontent.com/97817595/209577646-7ec2beb8-1d78-4b08-aa1e-a40dbe209725.png)
Side: I user the `User.register()` method, though the `passport-local-mongoose docs` have an async way of doing this as well.

But I had a pretty hard time trying to get the authentication to work since what I knew was a `req.login` method I used in Express.js that wasn't working at all here. So I needed to try something else. I went back to the `passport-local-mongoose docs` to check and see if there was something I missed and saw the async method I use in this next screenshot, and it worked perfectly! Per the [documentation](https://www.npmjs.com/package/passport-local-mongoose#asyncawait), when using the async method, there's no need to call `passport.serializeUser()` or `passport.deserializeUser()` prior to running the function. This is a screenshot of (`./pages/api/login.js`):
![login](https://user-images.githubusercontent.com/97817595/209578581-1dcbd413-5cbf-4ce7-b23f-10f74f235c53.png) 
Side 1: You may notice I start the login session in the login route but not the registration. That's only because I intend to have the registration api send a confirmation email, but I haven't built that function yet.
Side 2: If you're coming from using Express.js, you might also have used the `express-session` package, which isn't compatible with Next.js. For this I've found @hapi/iron to be a good replacement and would advise following the documentation outline provided by nextjs for its use. Here are the links for ease of reference: [`./lib/auth.js`](https://github.com/vercel/next.js/blob/canary/examples/with-passport/lib/auth.js) & [`./lib/auth-cookies.js`](https://github.com/vercel/next.js/blob/canary/examples/with-passport/lib/auth-cookies.js)

**Reply by SolwayEnterprises** [reply]
All that said, I might be able to help more specifically if I saw all that you're doing with your login process, but here are a couple guesses:
1. I notice that you're passing the user into the function via the function inputs and it's all lowercase, so I'm guessing it represents the username property rather than the UserSchema. If this is the case, you'll need to import the User from the UserSchema set up in `./models/User`.
2. Another reason it might not have worked for you is if you forgot to import and call `dbConnect()` prior to the function. I suggest this because I made that mistake myself.
